plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE TypeApplications   #-}
+{-# LANGUAGE TypeFamilies       #-}
+-- | Plutus implementation of an account that can be unlocked with a token.
+--   Whoever owns the token can spend the outputs locked by the contract.
+--   (A suitable token can be created with the 'Language.PlutusTx.Coordination.Contracts.Currency'
+--   contract, or with 'newAccount' in this module)
+module Language.PlutusTx.Coordination.Contracts.TokenAccount(
+  AccountOwner(..)
+  , pay
+  , redeem
+  , redeemTx
+  , balance
+  , address
+  , newAccount
+  ) where
+
+import           Control.Lens
+import qualified Data.Map                                          as Map
+import           Data.Maybe                                        (fromMaybe)
+
+import           Language.Plutus.Contract
+import qualified Language.PlutusTx                                 as PlutusTx
+
+import qualified Language.Plutus.Contract.Typed.Tx                 as TypedTx
+import           Ledger                                            (Address, PubKey, TxOut)
+import qualified Ledger                                            as Ledger
+import           Ledger.TxId                                       (TxId)
+import           Ledger.Typed.Scripts                              (ScriptType (..))
+import qualified Ledger.Typed.Scripts                              as Scripts
+import qualified Ledger.Validation                                 as V
+import           Ledger.Value                                      (CurrencySymbol, TokenName, Value)
+import qualified Ledger.Value                                      as Value
+
+import qualified Language.PlutusTx.Coordination.Contracts.Currency as Currency
+
+newtype AccountOwner = AccountOwner { unAccountOwner :: (CurrencySymbol, TokenName) }
+    deriving newtype Eq
+
+data TokenAccount
+
+instance ScriptType TokenAccount where
+    type RedeemerType TokenAccount = ()
+    type DataType TokenAccount = AccountOwner
+
+{-# INLINEABLE accountToken #-}
+accountToken :: AccountOwner -> Value
+accountToken (AccountOwner (symbol, name)) = Value.singleton symbol name 1
+
+{-# INLINEABLE validate #-}
+validate :: AccountOwner -> () -> V.PendingTx -> Bool
+validate owner _ ptx = V.valueSpent ptx `Value.geq` accountToken owner
+
+scriptInstance :: Scripts.ScriptInstance TokenAccount
+scriptInstance =
+    Scripts.Validator @TokenAccount $$(PlutusTx.compile [|| validate ||]) $$(PlutusTx.compile [|| wrap ||])
+    where
+    wrap = Scripts.wrapValidator @AccountOwner @()
+
+address :: Address
+address = Scripts.scriptAddress scriptInstance
+
+-- | Pay some money to the given token
+pay :: (AsContractError e, HasWriteTx s) => AccountOwner -> Value -> Contract s e TxId
+pay owner vl =
+    let ds = Ledger.DataScript (PlutusTx.toData owner)
+    in writeTxSuccess (payToScript vl address ds)
+
+ownsTxOut :: AccountOwner -> TxOut -> Bool
+ownsTxOut owner txout =
+    Just owner == (Ledger.txOutData txout >>= PlutusTx.fromData . Ledger.getDataScript)
```

Comment:
I'd bracket this a bit more - I don't think many people know what the relative precedence of `.` and `>>=` is...